File: python/semantic_kernel/filters/auto_function_invocation/auto_function_invocation_context.py
# Copyright (c) Microsoft. All rights reserved.

from typing import TYPE_CHECKING

from semantic_kernel.filters.filter_context_base import FilterContextBase

if TYPE_CHECKING:
    from semantic_kernel.connectors.ai.prompt_execution_settings import PromptExecutionSettings
    from semantic_kernel.contents.chat_history import ChatHistory
    from semantic_kernel.contents.function_call_content import FunctionCallContent
    from semantic_kernel.functions.function_result import FunctionResult


class AutoFunctionInvocationContext(FilterContextBase):
    """The context for auto function invocation filtering.

    This is the context supplied to the auto function invocation filters.

    Common use case are to alter the function_result, for instance filling it with a pre-computed
    value, in order to skip a step, for instance when doing caching.

    Another option is to terminate, this can be done by setting terminate to True.

    Args:
        function: The function invoked.
        kernel: The kernel used.
        arguments: The arguments used to call the function.
        is_streaming: Whether the function is streaming.
        chat_history: The chat history or None.
        function_call_content: The function call content or None.
        function_result: The function result or None.
        request_sequence_index: The request sequence index.
        function_sequence_index: The function sequence index.
        function_count: The function count.
        terminate: The flag to terminate.

    """

    chat_history: "ChatHistory | None" = None
    function_call_content: "FunctionCallContent | None" = None
    function_result: "FunctionResult | None" = None
    execution_settings: "PromptExecutionSettings | None" = None
    request_sequence_index: int = 0
    function_sequence_index: int = 0
    function_count: int = 0
    terminate: bool = False

File: nodejs/semantic-kernel/filters/auto-function-invocation/auto-function-invocation-context.ts
import { PromptExecutionSettings } from '../../connectors/ai/prompt-execution-settings'
import { ChatHistory } from '../../contents/chat-history'
import { FunctionCallContent } from '../../contents/function-call-content'
import { FunctionResult } from '../../functions/function-result'
import { KernelArguments } from '../../functions/kernel-arguments'
import { Kernel, KernelFunction } from '../../kernel'
import { FilterContextBase } from '../filter-context-base'

/**
 * Options for constructing an AutoFunctionInvocationContext.
 */
export interface AutoFunctionInvocationContextOptions {
  /** The function being invoked */
  function: KernelFunction
  /** The kernel instance */
  kernel: Kernel
  /** The arguments passed to the function */
  arguments: KernelArguments
  /** Whether the function is streaming */
  isStreaming?: boolean
  /** The chat history or null */
  chatHistory?: ChatHistory | null
  /** The function call content or null */
  functionCallContent?: FunctionCallContent | null
  /** The function result or null */
  functionResult?: FunctionResult | null
  /** The execution settings or null */
  executionSettings?: PromptExecutionSettings | null
  /** The request sequence index */
  requestSequenceIndex?: number
  /** The function sequence index */
  functionSequenceIndex?: number
  /** The function count */
  functionCount?: number
  /** The flag to terminate */
  terminate?: boolean
}

/**
 * The context for auto function invocation filtering.
 *
 * This is the context supplied to the auto function invocation filters.
 *
 * Common use case are to alter the function_result, for instance filling it with a pre-computed
 * value, in order to skip a step, for instance when doing caching.
 *
 * Another option is to terminate, this can be done by setting terminate to true.
 */
export class AutoFunctionInvocationContext extends FilterContextBase {
  /** The chat history or null */
  chatHistory: ChatHistory | null = null

  /** The function call content or null */
  functionCallContent: FunctionCallContent | null = null

  /** The function result or null */
  functionResult: FunctionResult | null = null

  /** The execution settings or null */
  executionSettings: PromptExecutionSettings | null = null

  /** The request sequence index */
  requestSequenceIndex: number = 0

  /** The function sequence index */
  functionSequenceIndex: number = 0

  /** The function count */
  functionCount: number = 0

  /** The flag to terminate */
  terminate: boolean = false

  constructor(options: AutoFunctionInvocationContextOptions) {
    super({
      function: options.function,
      kernel: options.kernel,
      arguments: options.arguments,
      isStreaming: options.isStreaming,
    })

    this.chatHistory = options.chatHistory ?? null
    this.functionCallContent = options.functionCallContent ?? null
    this.functionResult = options.functionResult ?? null
    this.executionSettings = options.executionSettings ?? null
    this.requestSequenceIndex = options.requestSequenceIndex ?? 0
    this.functionSequenceIndex = options.functionSequenceIndex ?? 0
    this.functionCount = options.functionCount ?? 0
    this.terminate = options.terminate ?? false
  }
}
